### PR TITLE
🐛 test: filter cluster-wide objects asserted in ResourceVersion tests to exclude objects of parallel tests

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

In a rare test case where k8s-upgrade-with-runtime-sdk and the quickstart tests's assertion on `ValidateResourceVersionStable` are run in-parallel, we can have a flaky test because:

Getting the owner graph included the RuntimeExtension. When `k8s-upgrade-with-runtime-sdk` finishes, it deleted the CR again, so the assertion in `ValidateResourceVersionStable` failed and the test failed.

xref: [failure occurency](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/10552/pull-cluster-api-e2e-main/1786405514213593088)

Also: enables the `ValidateResourceVersionStable` test for the `k8s-upgrade-with-runtime-sdk` to also test the ExtensionConfig.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/area e2e-testing

### CODE REVIEW

**Comment by chrischdi** [comment]
> @chrischdi: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cluster-api-apidiff-main	[4dbe1f7](https://github.com/kubernetes-sigs/cluster-api/commit/4dbe1f70498686afaee5a420e74689fba41f4258)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/10560/pull-cluster-api-apidiff-main/1787509420649877504)	false	`/test pull-cluster-api-apidiff-main`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api&pr=10560). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Achrischdi). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc) when you hit one in your PR.

This is fine, the changed function signature is explicitly marked not as stable, because of only being used in e2e tests.

**Comment by chrischdi** [comment]
> > @chrischdi: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> > Test name	Commit	Details	Required	Rerun command
> > pull-cluster-api-apidiff-main	[4dbe1f7](https://github.com/kubernetes-sigs/cluster-api/commit/4dbe1f70498686afaee5a420e74689fba41f4258)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/10560/pull-cluster-api-apidiff-main/1787509420649877504)	false	`/test pull-cluster-api-apidiff-main`
> > [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api&pr=10560). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Achrischdi). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc) when you hit one in your PR.
> 
> This is fine, the changed function signature is explicitly marked not as stable, because of only being used in e2e tests.

/override pull-cluster-api-apidiff-main

**Comment by chrischdi** [comment]
> @chrischdi: chrischdi unauthorized: /override is restricted to Repo administrators.

Let's ignore this, its optional anyway :D

File: test/framework/ownerreference_helpers.go [hunk 1]
```diff
@@ -382,6 +382,22 @@ func setClusterPause(ctx context.Context, cli client.Client, clusterKey types.Na
 	Expect(cli.Patch(ctx, cluster, pausePatch)).To(Succeed())
 }
 
+// GetOwnerGraphFilterByClusterNameFunc is used in e2e tests where the owner graph gets queried
+// to filter out cluster-wide objects which don't have the clusterName in their
+// object name. This avoids assertions on objects which are part of in-parallel
+// running tests like ExtensionConfig.
+func GetOwnerGraphFilterByClusterNameFunc(clusterName string) func(u unstructured.Unstructured) bool {
```
Comment (fabriziopandini) [inline test/framework/ownerreference_helpers.go:389]: what about:
SkipClusterObjectsWithoutNamePrefix

**Review by fabriziopandini — COMMENTED** [review]
just one nit (naming 😄 )

File: test/framework/ownerreference_helpers.go (hunk 1 above)
Comment (chrischdi) [reply test/framework/ownerreference_helpers.go:389]: SkipClusterObjectsWithoutNameContains

Because we don't match prefix

**Comment by chrischdi** [comment]
Note: #10530

was cherry-picked to v1.7 so we should fix at least the resourceVersion test in v1.7 too.

I suggest for v1.7 to just filter inside the new function to not change the function signatures there.

File: cmd/clusterctl/client/cluster/ownergraph.go [hunk 2]
```diff
@@ -39,11 +39,18 @@ type OwnerGraphNode struct {
 	Owners []metav1.OwnerReference
 }
 
+// GetOwnerGraphFilterFunction allows filtering the objects returned by GetOwnerGraph.
+// The function has to return true for objects which should be kept.
+// NOTE: this function signature is exposed to allow implementation of E2E tests; there is
+// no guarantee about the stability of this API. Using this test with providers may require
+// a custom implementation of this function, or the OwnerGraph it returns.
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/ownergraph.go:46]: This function doesn't return an OwnerGraph

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 3]
```diff
@@ -85,6 +85,10 @@ type clusterUpgradeWithRuntimeSDKSpecInput struct {
 	// Allows to inject a function to be run after test namespace is created.
 	// If not specified, this is a no-op.
 	PostNamespaceCreated func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace string)
+
+	// Allows to inject a function to be run after the cluster is upgraded.
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_runtimesdk.go:89]: Given how we usually use provisioned I think we should use a func name which matches the godoc more, e.g. what about "PostUpgrade"?

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 4]
```diff
@@ -260,6 +264,11 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
 			WaitForNodesReady: input.E2EConfig.GetIntervals(specName, "wait-nodes-ready"),
 		})
 
+		if input.PostMachinesProvisioned != nil {
+			log.Logf("Calling PostMachinesProvisioned for cluster %s", klog.KRef(namespace.Name, clusterResources.Cluster.Name))
+			input.PostMachinesProvisioned(input.BootstrapClusterProxy, namespace.Name, clusterName)
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_runtimesdk.go:269]: nit: Let's use the same clusterName for logging and to call the func

File: test/framework/ownerreference_helpers.go (hunk 1 above)
Comment (fabriziopandini) [reply test/framework/ownerreference_helpers.go]: SkipClusterObjectsWithNameFilter

File: test/framework/ownerreference_helpers.go [hunk 5]
```diff
@@ -382,6 +382,22 @@ func setClusterPause(ctx context.Context, cli client.Client, clusterKey types.Na
 	Expect(cli.Patch(ctx, cluster, pausePatch)).To(Succeed())
 }
 
+// SkipClusterObjectsWithoutNameContains is used in e2e tests where the owner graph
+// gets queried to filter out cluster-wide objects which don't have the clusterName
+// in their object name. This avoids assertions on objects which are part of in-parallel
+// running tests like ExtensionConfig.
+func SkipClusterObjectsWithoutNameContains(clusterName string) func(u unstructured.Unstructured) bool {
```
Comment (sbueringer) [inline test/framework/ownerreference_helpers.go:389]:
```suggestion
func FilterClusterObjectsWithoutNameContains(clusterName string) func(u unstructured.Unstructured) bool {
```

File: test/framework/ownerreference_helpers.go [hunk 6]
```diff
@@ -382,6 +382,22 @@ func setClusterPause(ctx context.Context, cli client.Client, clusterKey types.Na
 	Expect(cli.Patch(ctx, cluster, pausePatch)).To(Succeed())
 }
 
+// SkipClusterObjectsWithoutNameContains is used in e2e tests where the owner graph
+// gets queried to filter out cluster-wide objects which don't have the clusterName
+// in their object name. This avoids assertions on objects which are part of in-parallel
+// running tests like ExtensionConfig.
```
Comment (sbueringer) [inline test/framework/ownerreference_helpers.go:388]: Can we change clusterName to `s` or `str` (+update the godoc accordingly

(we can still mention it as an example)

File: test/framework/ownerreference_helpers.go [hunk 7]
```diff
@@ -382,6 +382,22 @@ func setClusterPause(ctx context.Context, cli client.Client, clusterKey types.Na
 	Expect(cli.Patch(ctx, cluster, pausePatch)).To(Succeed())
 }
 
+// SkipClusterObjectsWithoutNameContains is used in e2e tests where the owner graph
```
Comment (sbueringer) [inline test/framework/ownerreference_helpers.go:385]: I wonder if we should move this func to ownergraph.go

It seems to make a lot of sense there and here within "ownerreference_helpers.go" it fells a bit misplaced

**Comment by sbueringer** [comment]
> I suggest for v1.7 to just filter inside the new function to not change the function signatures there.

WDYT about checking via cs.k8s.io if someone even uses those funcs? Because I think if they don't, let's just keep it simple and backport this PR (CAPV can easily be adjusted)

**Comment by sbueringer** [comment]
Just checked. Only CAPI & CAPV. I would simply cherry-pick. We also have a disclaimer for our test packages regarding changes and it seems very unlikely that anyone is affected anyway.

File: test/e2e/cluster_upgrade_runtimesdk.go (hunk 3 above)
Comment (chrischdi) [reply test/e2e/cluster_upgrade_runtimesdk.go:89]: Renamed to PostUpgrade 👍

File: cmd/clusterctl/client/cluster/ownergraph.go (hunk 2 above)
Comment (chrischdi) [reply cmd/clusterctl/client/cluster/ownergraph.go]: Removed that part.

**Comment by sbueringer** [comment]
Thx

/lgtm

/assign @fabriziopandini

**Comment by fabriziopandini** [comment]
/lgtm
/approve

great job in catching this race condition while running E2E tests!

**Pull request merged**